### ISSUE TITLE
Enable http2 port option for serverless image

### ISSUE THREAD

**Issue by someone1** [cloudendpoints/esp#715]
I'm trying an (unsupported) deployment model where I run ESP on CloudRun for my gRPC Endpoints API and specify a backend pointed at a LoadBalancer for a Managed Instance Group of Windows VMs (this is to try and modernize a legacy component for the cloud, I can't containerize it)

I noticed the serverless image only uses the `--http-port` option, but gRPC requires the `--http2-port`. It's trivial to make my own image, but it would also be trivial to modify the script to pick the correct option based on an environmental variable so I wouldn't have to manage my own images with such a minor tweak.

Would you be open to adding this feature? It should be very easy/quick to do, I can even add a PR for it.

Out of curiosity, I didn't really find documentation for the `backend` definition in the gRPC yaml configuration file but know it [exists](https://cloud.google.com/service-infrastructure/docs/service-management/reference/rest/v1/services.configs#Backend). Is this feature just not documented or still in testing?

Additionally, the OpenAPI yaml configuration allows a `jwt_audience` value so ESP will sign forwarded requests to the backend enabling the use of IAP - I don't think the gRPC yaml configuration affords the same configuration. Is this by design? Maybe I'm about to quickly find out that IAP doesn't support gRPC but here's to hoping 🤞

**Comment by qiwzhang** [comment]
You are in the uncharted territory; all of them are not tested. I just want to point out a few things
1) for gRPC backend,   backend:address field can only has host name, not path, we don't do any path  translation.
2) By reading the code, if jwt_audience is specified, we do send id token.  But not sure if it works for grpc.
3) grpc backend proto is define [here](https://github.com/googleapis/googleapis/blob/master/google/api/backend.proto). It has jwt_audience field
4) For Cloud Run, you should open a http1 port specified by PORT environment,  Cloud run uses it for http health check,  you may need to open a separate port for http2 with another environment, you can submit a pr for it.

**Comment by someone1** [comment]
Hi and thanks for the quick response!

1. That's fine, I don't need it by path
2. I'll test and report back 👍 
3. Great! It's not documented in the API REST docs, does that mean its not available yet or just that the docs are out of date? It seems to have updated just fine.
4. I actually forked and modified the script to use http2 instead of http1 port and it doesn't report any issues though I'm now getting "service unavailable" responses.... I'll hack away at this but can I even specify more than 1 port in CloudRun?

**Comment by qiwzhang** [comment]
3. ESP code is using the proto defined in  [here](https://github.com/googleapis/googleapis/blob/master/google/api/backend.proto).  The doc may have not been updated.
4. Good question,  Cloud Run may only expose one port.  If you use it for http2, health check may fail.  You can use ssl_port which support both http1 and http2,  but not sure if Cloud run health check support https

**Comment by someone1** [comment]
### tl;dr

ESP can be used for gRPC APIs with custom  backends protected by IAP! Though I don't know if this can be done on CloudRun, and definitely cannot be done using the serverless image due to (seemingly erroneous) flag error messages

1. According to the [documentation](https://cloud.google.com/run/docs/issues) - Cloud Run supports gRPC hosted services except for gRPC streaming. So launching ESP for gRPC _should_ be possible but I'm not sure how. Is there any guidance you can provide on this?
2. Why can't the `--enable_backend_routing` flag  be used together with the `--grpc_backend_ssl flags` and/or  `--generate_self_signed_cert` flags?

### ESP on CloudRun

It doesn't appear I can use ESP on CloudRun as-is on anything but the HTTP1 port which is very discouraging. Using HTTP2 I get "Service Unavailable" and using HTTPS I get a nginx "400 Bad Request - The plain HTTP request was sent to HTTPS port" error response. 

According to the [documentation](https://cloud.google.com/run/docs/issues) - Cloud Run supports gRPC hosted services except for gRPC streaming. So launching ESP _should_ be possible but I'm not sure how. Is there any guidance you can provide on this?

I had the HTTP2 port working for a moment on CloudRun while testing - I'm not sure why it started giving me the "Service Unavailable" errors all of the sudden.

### Regarding the backend.jwt_audience + IAP: It works!

I cannot run the ESP serverless image on Compute Engine COS, I needed to use the standard image with the following flags: `--service=<service name> --ssl_port=8443 --backend=grpc://<address>:443 --generate_self_signed_cert --enable_grpc_backend_ssl`

Trying to get the serverless image working kept throwing errors:
`ERROR:Flag --enable_backend_routing cannot be used together with --grpc_backend_ssl flags.`

The same is true of the `--generate_self_signed_cert` flag. Why are these mutually exclusive?

Some tests:

* Connecting to LoadBalancer directly, IAP disabled: Works ✔️ 
* Connecting to ESP (using switches above), IAP disabled: Works ✔️  
  * From log outputs on my gRPC server, I can verify that an IAP compliant token is sent to the gRPC server from ESP when `jwt_audience` is set in the yaml configuration.
* Connecting to LoadBalancer directly, IAP enabled:  Fails ❌ 
  * client log output: `could not send request: rpc error: code = Internal desc = Bad Request: HTTP status code 400; transport: received the unexpected content-type "text/html; charset=UTF-8"`
  * Unauthenticated requests respond with 302
  * Unauthorized requests respond with 401
* Connecting to ESP (using switches above), IAP enabled: Works ✔️  

I'm not sure why the 3rd option failed, probably something with my local client tests.

**Comment by qiwzhang** [comment]
I am so glad to hear that you have made so much progress.

I have on clues how to turn on http2 port in Cloud Run.  You may have to ask Google Cloud run team.

My suggestion for ESP is not to use --enable_backend_routing and not to use serverless image.  just use standard image with yaml file with backend rules to set jwt_audiences.

**Comment by merlinnot** [comment]
Let me be a nice and friendly person and share a fully working configuration with you (Cloud Run on GKE/Cloud Run on Anthos).

```yaml
apiVersion: serving.knative.dev/v1alpha1
kind: Service
metadata:
  name: hello-esp
  namespace: default
spec:
  template:
    spec:
      containers:
        - args:
            [
              '--backend=grpc://grpc.api.my-company.com:80',
              '--http2_port=8080',
              '--rollout_strategy=managed',
              '--service_account_key=/mnt/credentials/service-account-credentials.json',
              '--service=hello.endpoints.my-project.cloud.goog',
            ]
          ports:
            - name: h2c
              containerPort: 8080
          image: gcr.io/endpoints-release/endpoints-runtime:1
          resources:
            limits:
              cpu: '1'
              memory: 256Mi
          volumeMounts:
            - mountPath: /mnt/credentials
              name: service-account-credentials
      serviceAccountName:
      containerConcurrency: 80
      timeoutSeconds: 300
      volumes:
        - name: service-account-credentials
          secret:
            secretName: service-account-credentials
  traffic:
    - latestRevision: true
      percent: 100
```

**Comment by merlinnot** [comment]
The only thing I need to fix here is to expose the actual server on internal network. But I think it should work just fine.

**Comment by someone1** [comment]
@merlinnot - Thanks! I'm using CloudRun Managed (not on GKE) so this is not an option for me. 

I'll try and test to see if I can get it running on Managed first, then contemplate introducing GKE into my application's design (was hoping to avoid this as we're a small team with no prior K8s experience)

**Comment by merlinnot** [comment]
Yeah, I’m also waiting for the managed one. My plan is to start with GKE and then move to managed once it’s feature-complete.

TBH it’s just a few clicks in the UI, so you can give it a shot.

Can you let me know if the above template helped you somehow in the managed environment? I don’t have time to try it myself (I need streaming anyway, which is not supported yet in managed), but I’d like to know.

**Comment by someone1** [comment]
@merlinnot - Sure, I just tried deploy making my own Dockerfile:

```Dockerfile
FROM gcr.io/endpoints-release/endpoints-runtime:1.40.0

ADD env-start-esp.py env-start-esp.py

ENTRYPOINT ["/env-start-esp.py"]
```
And the `env-start-esp.py` file copied from the serverless image, modified:
```python
#!/usr/bin/env python
#
# Copyright (C) Extensible Service Proxy Authors
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1. Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
# 2. Redistributions in binary form must reproduce the above copyright
#    notice, this list of conditions and the following disclaimer in the
#    documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
# OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
# HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
# OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
# SUCH DAMAGE.
#
###############################################################################
#

from __future__ import print_function
import os


def assert_env_var(name):
    if name not in os.environ:
        raise KeyError(
            "Serverless ESP expects {} in environment variables.".format(name)
        )


def make_error_app(error_msg):
    # error_msg must be a utf-8 or ascii bytestring
    def error_app(environ, start_response):
        start_response("503 Service Unavailable", [
                       ("Content-Type", "text/plain")])
        return [error_msg, "\n"]

    return error_app


def serve_error_msg(error_msg):
    print("Serving error handler with '{}'.".format(error_msg))
    import wsgiref.simple_server

    app = make_error_app(error_msg)
    port = int(os.environ["PORT"])
    server = wsgiref.simple_server.make_server("", port, app)
    server.serve_forever()


def main():
    CMD = "/usr/sbin/start_esp"
    PLATFORM = "Cloud Run"

    # The command being run must be the 0th arg.
    ARGS = [CMD, "--compute_platform_override='{}'".format(PLATFORM)]

    # Uncaught KeyError; if no port, we can't serve a nice error handler. Crash instead.
    assert_env_var("PORT")
    ARGS.append("--http2_port={}".format(os.environ["PORT"]))

    try:
        assert_env_var("ENDPOINTS_SERVICE_NAME")
    except KeyError as error:
        serve_error_msg(error.message)
    ARGS.append("--service={}".format(os.environ["ENDPOINTS_SERVICE_NAME"]))

    if "ENDPOINTS_SERVICE_VERSION" in os.environ:
        ARGS.extend(
            [
                "--rollout_strategy=fixed",
                "--version={}".format(os.environ["ENDPOINTS_SERVICE_VERSION"]),
            ]
        )
    else:
        ARGS.append("--rollout_strategy=managed")

    if "CORS_PRESET" in os.environ:
        ARGS.append("--cors_preset={}".format(os.environ["CORS_PRESET"]))

    if "ESP_ARGS" in os.environ:
        # By default, ESP_ARGS is comma-separated.
        # But if a comma needs to appear within an arg, there is an alternative
        # syntax: Pick a replacement delimiter, specify it at the beginning of the
        # string between two caret (^) symbols, and use it within the arg string.
        # Example:
        # ^++^--cors_allow_methods="GET,POST,PUT,OPTIONS"++--cors_allow_credentials
        arg_value = os.environ["ESP_ARGS"]

        delim = ","
        if arg_value.startswith("^") and "^" in arg_value[1:]:
            delim, arg_value = arg_value[1:].split("^", 1)
        if not delim:
            serve_error_msg("Malformed ESP_ARGS environment variable.")

        ARGS.extend(arg_value.split(delim))

    os.execv(CMD, ARGS)


if __name__ == "__main__":
    main()
```

Deployed to managed CloudRun with the following env vars:
`ENDPOINTS_SERVICE_NAME=<service name>`
`ESP_ARGS=--backend=grpc://<grpc api>:443,--enable_grpc_backend_ssl`


And its working! I'm not sure for how long though as I think the health checks will start failing soon and it will revert to "Service Unavailable" (although clearly it's working just fine)

**Comment by someone1** [comment]
@qiwzhang  - if this deployment stays stable - would you be open to modifying the serverless python script as follows:

* Introduce a check for env var ESP_USE_HTTP2=true to switch to using the `--http2_port` flag
* Detect if the `--backend` flag was passed into ESP_ARGS and drop the the `--enable_backend_routing` flag

That's all that appears necessary to get the current serverless image working for gRPC on CloudRun Managed (assuming my deployment stays stable for now)

**Comment by someone1** [comment]
Actually - the "service unavailable" only comes if I browse to the CloudRun URL via a browser, it works fine when connecting via a gRPC client!

**Comment by qiwzhang** [comment]
@someone1 I am open to a pr with http2 port.   But not enable_backend_route flag.
